zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
+	// sort vector
+	std::sort(vec.begin(), vec.end());
+
+	return (vec.size() != 0 ? vec[0] : -1); // return lowest value. Ideally, we'd want to throw out the bottom
+	// <some number>%, but that messes up depth calculation for things like spheres.
+	// If there are no values, return -1.
+
+	/* Test results (adaptive thresholding):
+	[ INFO] [1635948213.029497061]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal_8.jpg
+	[ INFO] [1635948213.689598629]: Calculated depth is 113.001 ✅
+	[ INFO] [1635948229.253951900]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal_behind_power_cell.png
+	[ INFO] [1635948229.870174338]: Calculated depth is 138.001 ✅
+	[ INFO] [1635948236.848476178]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_goal.png
+	[ INFO] [1635948237.445102845]: Calculated depth is 113.001 ✅
+	[ INFO] [1635948248.428153064]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_loading_bay.png
+	[ INFO] [1635948249.227361594]: Calculated depth is 141.001 ✅
+	[ INFO] [1635948255.684437760]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_low_goal.png
+	[ INFO] [1635948256.286312490]: Calculated depth is 92.001 ✅
+	[ INFO] [1635948266.234706198]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/cropped_power_cell.png
+	[ INFO] [1635948266.316279430]: Calculated depth is 0.001 ✅
+	[ INFO] [1635948273.702984310]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_behind_power_cell.png
+	[ INFO] [1635948275.086796499]: Calculated depth is 0.001 **non-cropped version** ❌
+	[ INFO] [1635948302.568242461]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/goal_with_power_cell_in_front.jpg
+	[ INFO] [1635948302.598558216]: Calculated depth is 0.001 ❌
+	[ INFO] [1635948312.950812216]: Received /home/ubuntu/2020Offseason/zebROS_ws/src/tf_object_detection/test_bitmaps/test_weirdness.jpg
+	[ INFO] [1635948312.980837687]: Calculated depth is 73.001 ✅
+	*/
+}
+
+// Get the most useful depth value in the cv::Mat depth contained within
+// the supplied bounding rectangle, using k-means
+double kMeansDepthMat(const cv::Mat& depth, const cv::Rect& bound_rect, bool debug = false, int maximumK = 3, float tolerance = 1e-3)
 {
-	double sum = 0.0;
-	unsigned count = 0;
-	for (int j = bound_rect.tl().y+1; j < bound_rect.br().y; j++) //for each row
-	{
-		const float *ptr_depth = depth.ptr<float>(j);
+	// setup randomizing (for initialization of k-means)
+	std::random_device seeder;
+	std::mt19937 engine(seeder());
+
+	// Start of checking if k is too high
 
-		for (int i = bound_rect.tl().x+1; i < bound_rect.br().x; i++) //for each pixel in row
+	// Convert cropped depth matrix to vector
```

Comment:
Perhaps a check similar to previous code for isContinouous Mats?

Reply:
What would accessing a non-existent value do? If it would return `nan`, the code uses `isnan` to check for that, but if not, I agree with adding that check.